### PR TITLE
HTTP POST Requests using AWS Signature Version 4

### LINKED ISSUE #212: V4 POST uploads
Presently this fails due to not parsing out and verifying the V4 signature.

### PR DESCRIPTION
**Pull request by michaelwittig**
I tried to not to refactor the code. The only thing I refactored it the way how the fields are checked by adding the `isField` method and some convenient method to calculate HmacSHA1 / HmacSHA256 because they share code.

I also added CORS headers.

closes #212

Let me know what you think.

### CODE REVIEW

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 1]
```diff
@@ -2811,6 +2845,40 @@ private static boolean startsWithIgnoreCase(String string, String prefix) {
...
+
+    private static byte[] hmacSHA1(byte[] data, byte[] key) {
+        return hmac("HmacSHA1", data, key);
+    }
+
+    private static byte[] hmacSHA256(byte[] data, byte[] key) {
+        return hmac("HmacSHA256", data, key);
+    }
+
+    private static String hex(byte[] bytes) {
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Prefer Guava `BaseEncoding.base16`.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 2]
```diff
@@ -2811,6 +2845,40 @@ private static boolean startsWithIgnoreCase(String string, String prefix) {
...
+        try {
+            Mac mac = Mac.getInstance(algorithm);
+            mac.init(new SecretKeySpec(key, algorithm));
+            return mac.doFinal(data);
+        } catch (InvalidKeyException | NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static byte[] hmacSHA1(byte[] data, byte[] key) {
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Remove this and the following wrapper.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java [hunk 3]
```diff
@@ -1757,68 +1758,97 @@ private void handlePostBlob(HttpServletRequest request,
...
         }
 
-        Map.Entry<String, BlobStore> provider =
-                blobStoreLocator.locateBlobStore(identity, null, null);
-        if (provider == null) {
-            response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+        String headerAuthorization = null;
+        S3AuthorizationHeader authHeader = null;
+        boolean signatureVersion4;
+        if (algorithm == null) {
```
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Do we need to check the signature configuration for v2, v4, or v2-or-v4 like we do for non-POST requests?

Comment (michaelwittig) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: I don't quiet understand what you mean by this. Can you point me to the lines in the code where such a check is performed?

**Comment by michaelwittig** [comment]
I adressed 2 of 3 of your comments. one thing left where I need some help.

File: src/main/java/org/gaul/s3proxy/S3ProxyHandler.java (hunk 3 above)
Comment (andrewgaul) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: Check out `switch (authHeader.authenticationType)` in `S3ProxyHandler.doHandle`.  S3Proxy allows configuration for V2-only, V4-only, and V2 or V4 which POST uploads should also enforce.

Comment (michaelwittig) [inline src/main/java/org/gaul/s3proxy/S3ProxyHandler.java]: done

**Comment by andrewgaul** [comment]
Thank you for your contribution @michaelwittig!

**Comment by michaelwittig** [comment]
@andrewgaul any plans to release 1.5.3 ?

**Comment by andrewgaul** [comment]
S3Proxy upgraded to Apache jclouds 2.0.2-SNAPSHOT so we must wait for its release.  I estimate a few months given jclouds release history.  I deployed S3Proxy 1.6.0-SNAPSHOT to Sonatype which you can use in the mean time:

https://github.com/andrewgaul/s3proxy/wiki/Using-S3Proxy-in-Java-projects

**Pull request merged**